datastore/importers/nmdb-import-systeminfo/Parser.test.cpp

```diff
@@ -0,0 +1,253 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+// =============================================================================
+// Maintained by Sandia National Laboratories <Netmeld@sandia.gov>
+// =============================================================================
+
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MAIN
+#include <boost/test/unit_test.hpp>
+#include <netmeld/datastore/parsers/ParserTestHelper.hpp>
+#include "Parser.hpp"
+
+namespace nmdp = netmeld::datastore::parsers;
+
+using qi::ascii::blank;
+
+class TestParser : public Parser
+{
+  public:
+    using Parser::start;
+    using Parser::hotfix;
+    using Parser::hotfixes;
+    using Parser::network_cards;
+    using Parser::network_card;
+    using Parser::ipAddressLine;
+};
+
+BOOST_AUTO_TEST_CASE(testNetworkCard)
+{
+  TestParser tp;
+  {
+    const auto& parserRule {tp.network_card};
+    std::vector<std::string> testsOk {
+      R"STR([01]: Realtek PCIe GbE Family Controller
+      Connection Name: Ethernet
+      DHCP Enabled:    Yes
+      IP address(es)
+      [01]: 192.168.0.2
+      [02]: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
+      )STR"
+    };
+
+    std::vector<std::string> testsFail {
+      R"STR(Connection Name: Ethernet
+      DHCP Enabled:    Yes
+      IP address(es)
+      [01]: 192.168.0.2
+      [02]: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
+      )STR", //Mising first string
+      R"STR([01]: Realtek PCIe GbE Family Controller
+      Connection Name:
+      DHCP Enabled:    Yes
+      IP address(es)
+      [01]: 192.168.0.2
+      [02]: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
+      )STR", //Missing connection name
+      R"STR([01]: Realtek PCIe GbE Family Controller
+      Connection Name: Ethernet
+      DHCP Enabled:
+      IP address(es)
+      [01]: 192.168.0.2
+      [02]: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
+      )STR", //Missing dhcp enabled data
+      R"STR([01]: Realtek PCIe GbE Family Controller
+      Connection Name: Ethernet
+      DHCP Enabled:    Yes
+      [01]: 192.168.0.2
+      [02]: 2001:0db8:85a3:0000:0000:8a2e:0370:7334
+      )STR", //Missing ipaddresses label
+      R"STR([01]: Realtek PCIe GbE Family Controller
+      Connection Name: Ethernet
+      DHCP Enabled:    Yes
+      IP address(es)
+      )STR" //Missing Ipaddress data
+    };
```

Comment:
My thought behind these was to ensure that the more complex rules actually get parsed? Is that a valid reason to keep them?

Reply:
I must be misunderstanding something then.  The object name is `testsFail` so I am assuming these are cases when the parser logic (i.e., `network_card`) should fail (and it is used as such in lines 100-103).  Thus, the text should not actually be consumed or other data objects populated.  So, I'm not understanding how it helps ensure the more complex rules actually get parsed.  I'm only seeing how it helps ensure a dev does not wrongly change the rule and succeed on something the parser should not.

Regardless, like I said it can be kept.  I was mainly pointing out there may be better alternatives depending on your objective and enumerating all the fail cases is not always needed.